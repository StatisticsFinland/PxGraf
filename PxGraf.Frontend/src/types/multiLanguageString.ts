/* istanbul ignore file */

/**
 * Type for string supporting multiple languages with each key representing a language and the value the string in that language.
 */
export type MultiLanguageString = { [key: string]: string };