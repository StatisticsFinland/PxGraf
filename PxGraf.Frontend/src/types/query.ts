/* istanbul ignore file */

/* Query - dimension queries*/

/**
 * Type for a query that contains @see {@link IDimensionQuery} objects for each dimension.
 * @property {{ [key: string]: IDimensionQuery }} - Queries for each dimension.
 */
export type Query = { [key: string]: IDimensionQuery };

/**
 * Enum that defines different dimension value filter types.
 */
export enum FilterType {
    Item = 'item',
    All = 'all',
    Top = 'top',
    From = 'from'
}

/**
 * Interface for dimension query properties.
 * @property {IValueFilter} valueFilter - Reference to value filter.
 * @property {boolean} selectable - Selected selectable dimension.
 * @property {any} virtualValueDefinitions - Virtual value definitions.
  */
export interface IDimensionQuery {
    valueFilter: IValueFilter
    selectable: boolean
    virtualValueDefinitions
}

/**
 * Interface for value filter properties.
 * @property {FilterType} type - Selected filter type.
 * @property {string[] | string | number} query - The query for the filter.
 */
export interface IValueFilter {
    type: FilterType;
    query?: string[] | string | number
}

/**
 * Type for string supporting multiple languages.
 */
type MultiLanguageString = { [key: string]: string };

/* CubeQuery - meta data editions*/

/**
 * Interface for cube query properties.
 * @property {MultiLanguageString} chartHeaderEdit - The chart header edition.
 * @property {{ [key: string]: IDimensionEditions }} variableQueries - The IDimensionEditions edition queries.
 */
export interface ICubeQuery {
    chartHeaderEdit?: MultiLanguageString;
    variableQueries: { [key: string]: IDimensionEditions };
}

/**
 * Interface for IDimensionEditions editions.
 * @property {{ [key: string]: IDimensionValueEditions }} valueEdits - The value editions.
 */
export interface IDimensionEditions {
    valueEdits: {
        [valueCode: string]: IDimensionValueEditions;
    };
}

/**
 * Interface for dimension value editions.
 * @property {MultiLanguageString} nameEdit - The name edition.
 * @property {IContentComponentEdition} contentComponent - The content component edition containing source and unit.
  */
export interface IDimensionValueEditions {
    nameEdit? : MultiLanguageString;
    contentComponent? : IContentComponentEdition;
}

/**
 * Interface for content component edition.
 * @property {MultiLanguageString} sourceEdit - Edition of the source.
 * @property {MultiLanguageString} unitEdit - Edition of the unit.
 */
interface IContentComponentEdition {
    sourceEdit? : MultiLanguageString;
    unitEdit? : MultiLanguageString;
}

/**
 * Interface that stores information about the query.
 * @property {number} size - The size of the query.
 * @property {number} sizeWarningLimit - The size warning limit of the query.
 * @property {number} maximumSupportedSize - The maximum supported size of the query.
 * @property {string[]} validVisualizations - Names of valid visualization types for of the query.
 * @property {{[key: string]: MultiLanguageString}} visualizationRejectionReasons - Reasons for why certain visualization types are rejected for the query.
 * @property {number} maximumHeaderLength - The maximum header length of the query.
 */
export interface IQueryInfo {
    size: number;
    sizeWarningLimit: number;
    maximumSupportedSize: number;
    validVisualizations: string[];
    visualizationRejectionReasons: {[key: string]: MultiLanguageString};
    maximumHeaderLength: number;
}