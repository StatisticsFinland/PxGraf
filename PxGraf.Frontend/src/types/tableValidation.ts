/**
 * Interface for the result of table meta data validation
 * @property {boolean} tableHasContentDimension - True if the table has a content dimension
 * @property {boolean} tableHasTimeDimension - True if the table has a time dimension
 * @property {boolean} allDimensionsContainValues - True if all dimensions contain values
 */
export interface ITableMetaValidationResult {
    tableHasContentDimension: boolean,
    tableHasTimeDimension: boolean,
    allDimensionsContainValues: boolean,
}