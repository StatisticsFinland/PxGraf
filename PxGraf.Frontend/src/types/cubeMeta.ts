/* istanbul ignore file */

import { MultiLanguageString } from "./multiLanguageString";

/**
 * Interface for cube meta properties. A cube is a multi-dimensional result of Px file data filtered by a query.
 * @property {string[]} languages - List of available languages.
 * @property {MultiLanguageString} header - The multi language header of the cube.
 * @property {MultiLanguageString} note - The note of the cube.
 * @property {IVariable[]} variables - List of variables.
 */
export interface IMatrixMetadata {
    DefaultLanguage: string,
    AvailableLanguages: string[],
    Dimensions: IDimension[],
    AdditionalProperties?: { [key: string]: IMetaProperty }
}

/**
 * TODO: Summary
 */
export interface IMetaProperty {
    KeyWord: string,
    CanGetStringValue: boolean,
    CanGetMultilanguageValue: boolean,
    Entries: MultiLanguageString | string
}

/**
 * Interface for dimnension properties. Each variable defines a dimension in a cube.
 * @property {string} code - The code of the variable.
 * @property {VariableType} type - The type of the variable.
 * @property {MultiLanguageString} name - The multi language name of the variable.
 * @property {{ [key: string]: IMetaProperty }} additionalProperties - Optional dictionary for additional metadata properties
 * @property {IVariableValue[]} values - List of values associated with this variable.
 */
export interface IDimension {
    Code: string,
    Name: MultiLanguageString,
    AdditionalProperties?: { [key: string]: IMetaProperty },
    Values: IDimensionValue[]
    Type: VariableType,
}

/**
 * Interface for variable properties.Each variable defines a dimension in a cube.
 * @property { string } code - The code of the variable.
 * @property { MultiLanguageString } name - The multi language name of the variable.
 * @property { MultiLanguageString } note - The multi language note of the variable.
 * @property { VariableType } type - The type of the variable.
 * @property { IVariableValue[] } values - List of values associated with this variable.
 */
export interface IVariable {
    code: string,
    name: MultiLanguageString,
    note: MultiLanguageString,
    type: VariableType,
    values: IVariableValue[]
}

/**
 * Interface for dimension value properties.
 * @property {string} code - The code of the value.
 * @property {MultiLanguageString} name - The multi language name of the value.
 * @property {{ [key: string]: IMetaProperty }} additionalProperties - Optional dictionary for additional metadata properties
 */
export interface IDimensionValue {
    Code: string,
    Name: MultiLanguageString,
    Virtual: boolean
    AdditionalProperties?: { [key: string]: IMetaProperty },
}

/**
 * Interface for variable value properties.
 * @property { string } code - The code of the value.
 * @property { MultiLanguageString } name - The multi language name of the value.
 * @property { MultiLanguageString } note - The multi language note of the value.
 * @property { boolean } isSum - Flag to indicate if the value is an elimination value.
 * @property { IContentComponent } contentComponent - The content variable component of the value.
 * @property { IVirtualComponent[] } virtualComponents - List of virtual components.
 */
export interface IVariableValue {
    code: string,
    name: MultiLanguageString,
    note: MultiLanguageString,
    isSum: boolean,
    contentComponent?: IContentComponent
    virtualComponents?: IVirtualComponent[]
}

/**
 * Interface for value content variable value properties.
 * @property {MultiLanguageString} unit - The unit of the value.
 * @property {MultiLanguageString} source - The source of the value.
 * @property {number} numberOfDecimals - The precision, number of decimals of the value.
 * @property {string} lastUpdated - The last updated date of the value.
 */
export interface IContentComponent {
    unit: MultiLanguageString,
    source: MultiLanguageString,
    numberOfDecimals: number,
    lastUpdated: string
}

/**
 * Interface for virtual component properties.
 * @property {string} operator - The operator of the virtual component.
 * @property {string[]} operandCoded - List of coded operands.
 */
export interface IVirtualComponent {
    operator: string,
    operandCoded: string[]
}

/**
 * Enumeration for variable types.
 */
export enum VariableType {
    Unknown = 'N',
    Time = 'T',
    Ordinal = 'P',
    Geographical = 'G',
    Content = 'C',
    OtherClassificatory = 'F',
}

export function getVariablesFromMatrixMetadata(meta: IMatrixMetadata): IVariable[] {
    return meta.Dimensions.map(d => {
        return {
            code: d.Code,
            name: d.Name,
            note: d.Name,
            type: d.Type,
            values: d.Values.map(v => convertDimensionValueToVariableValue(v, meta))
        }
    })
}

export function convertDimensionValueToVariableValue(value: IDimensionValue, meta: IMatrixMetadata): IVariableValue {
    const dimension: IDimension = meta.Dimensions.find(d => d.Values.find(v => v.Code === value.Code));
    const contentComponent = (): IContentComponent | null => {
        if (dimension.Type === VariableType.Content) {
            return {
                unit: getAdditionalProperty("UNIT", value.AdditionalProperties) as MultiLanguageString,
                source: getAdditionalProperty("SOURCE", value.AdditionalProperties) as MultiLanguageString,
                numberOfDecimals: parseInt(getAdditionalProperty("PRECISION", value.AdditionalProperties, true) as string),
                lastUpdated: getAdditionalProperty("LAST-UPDATED", value.AdditionalProperties, true) as string
            }
        }
        else return null;
    }
    return {
        code: value.Code,
        name: value.Name,
        note: value.Name,
        isSum: getValueIsSumValue(value, dimension),
        contentComponent: contentComponent()
    }
}

export function getValueIsSumValue(value: IDimensionValue, dimension: IDimension): boolean {
    return getAdditionalProperty("ELIMINATION", dimension.AdditionalProperties) === value.Code;
}

function getAdditionalProperty(key: string, additionalProperties: { [key: string]: IMetaProperty }, forceToString = false): string | MultiLanguageString {
    if (additionalProperties[key]) {
        if (additionalProperties[key].CanGetStringValue) {
            return additionalProperties[key].Entries[0];
        }
        else if (additionalProperties[key].CanGetMultilanguageValue) {
            if (forceToString) return additionalProperties[key].Entries[0];
            else return additionalProperties[key].Entries;
        }
    }
    return null;
}
