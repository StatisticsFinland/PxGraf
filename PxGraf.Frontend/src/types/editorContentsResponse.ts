import { MultiLanguageString } from "./multiLanguageString";
import { VisualizationType } from "./visualizationType";

/**
 * Interface for dimension sorting options.
 * @property {string} code - The code of the sorting option.
 * @property {MultiLanguageString} description - Longer, user facing description of the sorting option.
 */
export interface ISortingOption {
    code: string;
    description: MultiLanguageString;
}

/**
 * Interface for sorting option collection. Contains default and pivoted sorting options if available.
 * @property {ISortingOption[]} default - Default sorting options.
 * @property {ISortingOption[]} pivoted - Pivoted sorting options.
 */
interface ISortingOptionCollection {
    default: ISortingOption[];
    pivoted: ISortingOption[];
}

/**
 * Interface for visualization options for a specific visualization type.
 * @property {VisualizationType} type - The type of visualization.
 * @property {boolean} [allowShowingDataPoints] - Flag to indicate if showing data points is allowed.
 * @property {boolean} [allowCuttingYAxis] - Flag to indicate if cutting the Y-axis is allowed.
 * @property {boolean} [allowMatchXLabelsToEnd] - Flag to indicate if matching X labels to the end is allowed.
 * @property {boolean} [allowSetMarkerScale] - Flag to indicate if setting marker scale is allowed.
 * @property {boolean} allowManualPivot - Flag to indicate if manual pivoting is allowed.
 * @property {boolean} allowMultiselect - Flag to indicate if multivalue selectable is allowed.
 * @property {ISortingOptionCollection} sortingOptions - Sorting options for the visualization containing default and pivoted options.
 */
export interface IVisualizationOptions {
    type: VisualizationType;
    allowShowingDataPoints?: boolean;
    allowCuttingYAxis?: boolean;
    allowMatchXLabelsToEnd?: boolean;
    allowSetMarkerScale?: boolean;
    allowManualPivot: boolean;
    allowMultiselect: boolean;
    sortingOptions: ISortingOptionCollection;
}

/**
 * Interface for the response from the editor contents API used to populate the settings for adjusting the visualization in the editor.
 * @property {number} size - The size of the response based on the amount of values selected from all dimensions.
 * @property {number} sizeWarningLimit - The size warning limit for the response.
 * @property {number} maximumSupportedSize - The maximum supported size for the response.
 * @property {MultiLanguageString} headerText - Default header text for the visualization as a @ {@link MultiLanguageString} object.
 * @property {number} maximumHeaderLength - The maximum length of the header text.
 * @property {IVisualizationOptions[]} visualizationOptions - Visualization options for the valid visualization types as an array of @ {@link IVisualizationOptions} objects.
 * @property {object} visualizationRejectionReasons - Object containing rejection reasons for each visualization type as a @ {@link MultiLanguageString} object.
 */
export interface IEditorContentsResponse {
    size: number;
    sizeWarningLimit: number;
    maximumSupportedSize: number;
    headerText: MultiLanguageString;
    maximumHeaderLength: number;
    visualizationOptions: IVisualizationOptions[];
    visualizationRejectionReasons: {
        [visualizationType: string]: MultiLanguageString;
    }
}