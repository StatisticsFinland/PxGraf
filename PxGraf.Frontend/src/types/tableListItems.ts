/* istanbul ignore file */

import { MultiLanguageString } from "./multiLanguageString";

/**
 * Interface for a section in the hierarchy of the table list.
 * @property {IDatabaseGroupHeader[]} headers - The headers in the section.
 * @property {IDatabaseTable[]} files - The tables in the section.
 */
export interface IDatabaseGroupContents {
    headers: IDatabaseGroupHeader[];
    files: IDatabaseTable[];
}

/**
 * Interface for items shown on the table listing that can be sorted by the translated name.
 * @property {MultiLanguageString} name - Localized names in all available languages.
 * @property {string[]} languages - Available languages of the item. 
 */
export interface ISortableTableListItem {
    name: MultiLanguageString;
    languages: string[];
}

/***
 * Interface for a table group header
 * @property {string} code - The code of the table group.
 * @property {MultiLanguageString} name - The name of the table group as a multi-language string.
 */
export interface IDatabaseGroupHeader extends ISortableTableListItem {
    code: string;
}

/***
 * Enum for database table errors
 */
export enum EDatabaseTableError {
    contentLoad = 'ContentLoad',
    contentDimensionMissing = 'ContentDimensionMissing',
    timeDimensionMissing = 'TimeDimensionMissing',
}

/***
 * Interface for a px table
 * @property {fileName} fileName - Name of the px table file.
 * @property {string} lastUpdated - The last updated date of the table.
 * @property {EDatabaseTableError} error - Enum to indicate if there's an error with the table.
 */
export interface IDatabaseTable extends ISortableTableListItem {
    fileName: string;
    lastUpdated: string | null;
    error?: EDatabaseTableError;
}
