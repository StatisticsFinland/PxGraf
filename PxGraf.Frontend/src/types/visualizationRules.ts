/* istanbul ignore file */

import { MultiLanguageString } from "./multiLanguageString"

/**
 * Interface for visualization rules.
 * @property allowManualPivot - Flag to indicate if manual pivot is allowed.
 * @property sortingOptions - List of sorting options.
 * @property multiselectVariableAllowed - Flag to indicate if multiselect variable is allowed.
 * @property visualizationTypeSpecificRules - Object for visualization type specific rules.
 */
export interface IVisualizationRules {
    allowManualPivot: boolean,
    sortingOptions: ISortingOption[],
    multiselectVariableAllowed: boolean
    visualizationTypeSpecificRules?: ITypeSpecificVisualizationRules
}

/**
 * Interface for different sorting options
 * @property code - Code for the sorting option
 * @property description - Description for the sorting option as a multi language string
 */
export interface ISortingOption {
    code: string,
    description: MultiLanguageString
}

/**
 * Interface for visualization type specific rules.
 * @property allowShowingDataPoints - Flag to indicate if showing data points is allowed.
 * @property allowCuttingYAxis - Flag to indicate if cutting the Y-axis is allowed.
 * @property allowMatchXLabelsToEnd - Flag to indicate if matching X labels to the end is allowed.
 * @property allowSetMarkerScale - Flag to indicate if setting marker scale is allowed.
 */
export interface ITypeSpecificVisualizationRules {
    allowShowingDataPoints: boolean,
    allowCuttingYAxis: boolean,
    allowMatchXLabelsToEnd: boolean,
    allowSetMarkerScale: boolean
}