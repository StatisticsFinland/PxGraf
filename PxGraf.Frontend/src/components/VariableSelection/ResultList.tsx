import React from 'react';
import { useTranslation } from 'react-i18next';
import { Stack, List, ListItem, ListItemText, Typography, Skeleton } from '@mui/material';
import styled from 'styled-components';
import { IDimensionValue } from 'types/cubeMeta';
import { UiLanguageContext } from 'contexts/uiLanguageContext';

interface IResultListProps {
    variableValues: IDimensionValue[]
    resolvedVariableValueCodes: string[]
}

const DenseList = styled(List)`
    width: 100%;
    border-style: solid;
    border-width: 1px;
    border-radius: 4px;
    position: relative;
    overflow: auto;
    max-height: 200px;
`;

const NoPaddingListItem = styled(ListItem)`
    padding-top: 0;
    padding-bottom: 0;
`;

const StyledEm = styled.em`
    color: #ccc;
`;

const StyledSkeleton = styled(Skeleton)`
    width: 50%;
`;

const DenseListWrapper = styled(Stack)`
    width: 100%;
`;

export const ResultList: React.FC<IResultListProps> = ({ variableValues, resolvedVariableValueCodes }) => {
    const { t } = useTranslation();
    const { uiContentLanguage } = React.useContext(UiLanguageContext);

    let listContent;
    if (resolvedVariableValueCodes == null) {
        listContent =
            <NoPaddingListItem>
                <StyledSkeleton variant="text" />
            </NoPaddingListItem>
    }
    else if (resolvedVariableValueCodes.length === 0) {
        listContent =
            <NoPaddingListItem>
                <StyledEm>{t("general.noResults")}</StyledEm>
            </NoPaddingListItem>
    }
    else {
        const resultTexts = resolvedVariableValueCodes.map(valueCode => {
            const value = variableValues.find(value => value.Code === valueCode);
            return value?.Name[uiContentLanguage] ?? valueCode;
        });

        listContent = resultTexts.map((value) =>
            <NoPaddingListItem key={value}>
                <ListItemText primary={value} />
            </NoPaddingListItem>
        );
    }

    return (
        <DenseListWrapper>
            <Typography>{t("general.results")}:</Typography>
            <DenseList dense
                sx={{
                    bgcolor: 'background.paper',
                    borderColor: 'divider',
                    '& ul': { padding: 0 },
                }}
            >
                {listContent}
            </DenseList>
        </DenseListWrapper>
    );
}

export default ResultList;