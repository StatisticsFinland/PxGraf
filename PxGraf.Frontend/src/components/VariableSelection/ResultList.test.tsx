import React from 'react';
import { render } from "@testing-library/react";
import { IDimensionValue } from "../../types/cubeMeta";
import ResultList from "./ResultList";

const mockDimensionValues: IDimensionValue[] = [
    {
        code: "2018",
        name: {
            fi: "2018",
            sv: "2018",
            en: "2018"
        },
        isVirtual: false,
    },
    {
        code: "2019",
        name: {
            fi: "2019",
            sv: "2019",
            en: "2019"
        },
        isVirtual: false,
    },
    {
        code: "2020",
        name: {
            fi: "2020",
            sv: "2020",
            en: "2020"
        },
        isVirtual: false,
    },
    {
        code: "2021",
        name: {
            fi: "2021*",
            sv: "2021*",
            en: "2021*"
        },
        isVirtual: false,
    }
];

jest.mock('react-i18next', () => ({
    ...jest.requireActual('react-i18next'),
    useTranslation: () => {
        return {
            t: (str: string) => str,
            i18n: {
                changeLanguage: () => new Promise(() => null),
            },
        };
    },
}));

describe('Rendering test', () => {
    it('renders correctly', () => {
        const { asFragment } = render(<ResultList
            dimensionValues={mockDimensionValues}
            resolvedDimensionValueCodes={["2019", "2020"]}
        ></ResultList>);
        expect(asFragment()).toMatchSnapshot();
    });
});