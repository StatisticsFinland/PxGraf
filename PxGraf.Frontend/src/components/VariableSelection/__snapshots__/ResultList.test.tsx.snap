// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering test renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiStack-root sc-dntaoT eQwKBU css-nen11g-MuiStack-root"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
    >
      general.results:
    </p>
    <ul
      class="MuiList-root MuiList-padding MuiList-dense sc-blHHSb ckywJh css-1f8gdww-MuiList-root"
    >
      <li
        class="MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-padding sc-gtLWhw iKRme css-6lnq02-MuiListItem-root"
      >
        <div
          class="MuiListItemText-root MuiListItemText-dense css-tlelie-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
          >
            2019
          </span>
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-padding sc-gtLWhw iKRme css-6lnq02-MuiListItem-root"
      >
        <div
          class="MuiListItemText-root MuiListItemText-dense css-tlelie-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
          >
            2020
          </span>
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
